### ISSUE TITLE
Crates can drop several items from the main loot pools at once, instead of only one

### ISSUE THREAD

**Issue by TerrorPenguin** [bluemagic123/tModLoader#4050]
### Version

Latest 1.4.4-stable

### OS

Windows

### Platform

Steam

### This bug affects...

Gameplay as a Player

### Description

The way that vanilla handles crate loot drops is inside a while (flag) loop, and when any of the primary loot pools are selected, it sets the flag to false.
This means, for example, it's impossible to get both sailfish boots AND ore from a single crate.
Some items such as healing/mana potions and bait are excluded from this in vanilla, and can drop alongside other items.

tModLoader on the other hand, uses the "AlwaysAtLeastOneSuccess" drop rule, which allows for any of the main pools to drop simultaneously.
I'm not completely familiar with the ItemDropRule stuff, so I'm not sure what the immediate solution should be.

The following image is a screenshot of the loot from a single hellstone crate (this should be impossible).
![image](https://github.com/tModLoader/tModLoader/assets/41289369/b5d40bca-32f9-43ac-8cc4-9e8e8595c76f)

### Steps to reproduce

Open any crate until you get several primary items

### Expected Behavior

Only one primary item should be selected for each crate

### Actual Behavior

A random amount of primary items from 1-X is selected

### Reproduction frequency

100%

**Comment by JavidPack** [comment]
> The way that vanilla handles crate loot drops is inside a while (flag) loop, and when any of the primary loot pools are selected, it sets the flag to false.
> This means, for example, it's impossible to get both sailfish boots AND ore from a single crate.

The ore drop is actually not gated by a check for that flag but it does set that flag. Therefore it is possible to get a primary loot AND an ore if they both succeed in the same while loop iteration. This is a vanilla Terraria result I found in my testing showing a primary loot and an ore:
![image](https://github.com/tModLoader/tModLoader/assets/4522492/b848c691-9127-4bd5-8a37-331db5d2851c)

As for your hellstone crate results, the way tModLoader is using the `AlwaysAtLeastOneSuccess` rule does not allow multiple primary loot to drop because the primary loot are within a nested separate `SequentialRulesNotScalingWithLuck` rule that only allows up to 1 success.

I was not able to replicate your hellstone crate results. It shouldn't be possible to get multiple ores or multiple bars or multiple primary loot. Since you didn't upload your client.log file, I can only assume you had a mod enabled while doing this testing. 

Please disable all mods and attempt to replicate that hellstone crate behavior.

**Comment by TerrorPenguin** [comment]
This was my mistake, I didn't realise while loops continued to the end without a breakpoint even if the conditions for it were no longer true.
The several bars/ores is a separate issue, I should've taken a screenshot with mods disabled rather than using that screenshot.

**Comment by JavidPack** [comment]
Ok, so can you replicate this with 0 mods, or should we mark this as solved?
